android/expoview/src/main/java/versioned/host/exp/exponent/modules/universal/notifications/ScopedExpoNotificationPresentationModule.kt

```diff
@@ -0,0 +1,133 @@
+package versioned.host.exp.exponent.modules.universal.notifications
+
+import android.content.Context
+import android.os.Bundle
+import android.os.ResultReceiver
+import expo.modules.core.Promise
+import expo.modules.notifications.notifications.NotificationSerializer
+import expo.modules.notifications.notifications.interfaces.NotificationTrigger
+import expo.modules.notifications.notifications.model.Notification
+import expo.modules.notifications.notifications.model.NotificationContent
+import expo.modules.notifications.notifications.model.NotificationRequest
+import expo.modules.notifications.notifications.presentation.ExpoNotificationPresentationModule
+import expo.modules.notifications.service.NotificationsService
+import host.exp.exponent.kernel.ExperienceKey
+import host.exp.exponent.notifications.ScopedNotificationsUtils
+import host.exp.exponent.notifications.model.ScopedNotificationRequest
+import java.util.*
+
+class ScopedExpoNotificationPresentationModule(
+  context: Context,
+  private val experienceKey: ExperienceKey
+) : ExpoNotificationPresentationModule(context) {
+  private val scopedNotificationsUtils: ScopedNotificationsUtils = ScopedNotificationsUtils(context)
+
+  override fun createNotificationRequest(
+    identifier: String,
+    content: NotificationContent,
+    trigger: NotificationTrigger?
+  ): NotificationRequest {
+    return ScopedNotificationRequest(identifier, content, trigger, experienceKey.scopeKey)
+  }
+
+  override fun serializeNotifications(notifications: Collection<Notification>): ArrayList<Bundle> {
+    val serializedNotifications = arrayListOf<Bundle>()
+    for (notification in notifications) {
+      if (scopedNotificationsUtils.shouldHandleNotification(notification, experienceKey)) {
+        serializedNotifications.add(NotificationSerializer.toBundle(notification))
+      }
+    }
+    return serializedNotifications
+  }
+
+  override fun dismissNotificationAsync(identifier: String, promise: Promise) {
+    NotificationsService.getAllPresented(
+      context,
+      object : ResultReceiver(null) {
+        override fun onReceiveResult(resultCode: Int, resultData: Bundle) {
+          super.onReceiveResult(resultCode, resultData)
+          val notifications = resultData.getParcelableArrayList<Notification>(
+            NotificationsService.NOTIFICATIONS_KEY
+          )
+          if (resultCode == NotificationsService.SUCCESS_CODE && notifications != null) {
+            val notification = findNotification(notifications, identifier)
+            if (notification == null || !scopedNotificationsUtils.shouldHandleNotification(notification, experienceKey)) {
+              promise.resolve(null)
+              return
+            }
+            doDismissNotificationAsync(identifier, promise)
+          } else {
+            val e = resultData.getSerializable(NotificationsService.EXCEPTION_KEY) as Exception
```

Comment:
What happened with getParcelable?

Reply:
Great question!

`resultData.getParcelable<Exception>(NotificationsService.EXCEPTION_KEY)` doesn't compile since `Exception` doesn't extend `Parcelable`.

So, I looked up all the callsites where this is set (found usages of EXCEPTION_KEY) and found that the setter of this uses `putSerializable`, I'm assuming for the same reason: https://github.com/expo/expo/blob/master/packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/NotificationsService.kt#L606

Thus, I changed it to getSerializable.